File: ios/NorthAx/Services/IntervalsService.swift
import Foundation
import Observation
import AuthenticationServices
#if canImport(UIKit)
import UIKit
#endif

/// Connects to intervals.icu (OAuth) — the man-in-the-middle over Garmin/Strava
/// (§9): the client secret never touches the app. `connect()` opens the
/// backend-issued authorization URL in an `ASWebAuthenticationSession`; on the
/// `northax://intervals/connected` callback the backend has stored the tokens, so
/// the app polls status and triggers a sync.
@MainActor
@Observable
class IntervalsService {
    var connectionState: IntervalsConnectionState = .disconnected
    var syncedActivities: [GarminActivity] = []
    var isSyncing: Bool = false

    private let api = NorthAxAPI.shared
    private let contextProvider = WebAuthContextProvider()
    private var authSession: ASWebAuthenticationSession?

    // MARK: - Status

    /// Load the current connection state from the backend (call on app start).
    func refreshStatus() async {
        if let state = try? await api.intervalsStatus() {
            connectionState = state
            if state.isConnected {
                syncedActivities = (try? await api.activities(limit: 30)) ?? syncedActivities
            }
        }
    }

    // MARK: - Connection

    func connect() async {
        connectionState = .connecting
        do {
            let url = try await api.intervalsAuthorizationURL()
            try await presentAuth(url: url)
            connectionState = try await api.intervalsStatus()
            await syncActivities()
        } catch let error as APIError {
            if error.isIntervalsNotConfigured {
                connectionState = .error("intervals.icu isn’t enabled on this server yet. Please contact support.")
            } else {
                connectionState = .error(error.userMessage)
            }
        } catch {
            connectionState = .error("Couldn’t connect to intervals.icu.")
        }
    }

    /// Connect with a personal intervals.icu API key (no web OAuth).
    func connectWithAPIKey(athleteId: String, apiKey: String) async {
        connectionState = .connecting
        do {
            connectionState = try await api.connectWithAPIKey(athleteId: athleteId, apiKey: apiKey)
            await syncActivities()
        } catch let error as APIError {
            connectionState = .error(error.userMessage)
        } catch {
            connectionState = .error("Couldn't connect with that API key.")
        }
    }

    func disconnect() {
        Task {
            try? await api.intervalsDisconnect()
            connectionState = .disconnected
            syncedActivities = []
        }
    }

    // MARK: - Sync

    func syncActivities() async {
        guard connectionState.isConnected else { return }
        isSyncing = true
        defer { isSyncing = false }
        do {
            try await api.intervalsSync()
            syncedActivities = try await api.activities(limit: 30)
            connectionState = try await api.intervalsStatus()
        } catch {
            // Keep the last good state; a transient sync failure isn't fatal.
        }
    }

    // MARK: - Plan push (§9.4)

    /// Push a planned session to intervals.icu (→ Garmin) as a scheduled workout.
    /// Returns whether the push succeeded.
    @discardableResult
    func pushPlannedSession(_ session: PlannedSession, on date: Date) async -> Bool {
        guard connectionState.isConnected else { return false }
        return (try? await api.pushWorkout(date: date, session: session)) != nil
    }

    // MARK: - ASWebAuthenticationSession

    private func presentAuth(url: URL) async throws {
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            let session = ASWebAuthenticationSession(
                url: url, callbackURLScheme: APIConfig.appScheme
            ) { _, error in
                if let error {
                    // User-cancelled is not a hard error.
                    if (error as? ASWebAuthenticationSessionError)?.code == .canceledLogin {
                        continuation.resume(throwing: CancellationError())
                    } else {
                        continuation.resume(throwing: error)
                    }
                } else {
                    continuation.resume()
                }
            }
            session.presentationContextProvider = contextProvider
            session.prefersEphemeralWebBrowserSession = false
            self.authSession = session
            if !session.start() {
                continuation.resume(throwing: APIError.offline)
            }
        }
    }
}

/// Supplies the window to anchor the auth sheet to.
final class WebAuthContextProvider: NSObject, ASWebAuthenticationPresentationContextProviding {
    func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        #if canImport(UIKit)
        let scene = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first { $0.activationState == .foregroundActive }
        return scene?.keyWindow ?? ASPresentationAnchor()
        #else
        return ASPresentationAnchor()
        #endif
    }
}

File: android/app/src/main/java/app/northax/store/IntervalsService.kt
package app.northax.store

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import app.northax.data.remote.ApiError
import app.northax.data.remote.NorthAxApi
import app.northax.domain.model.GarminActivity
import app.northax.domain.model.IntervalsConnectionState
import app.northax.domain.model.PlannedSession
import java.time.LocalDate

/**
 * Connects to intervals.icu (OAuth) — the man-in-the-middle over Garmin/Strava:
 * the client secret never touches the app. `beginConnect()` returns the
 * backend-issued authorization URL for a Custom Tab; on the
 * `northax://intervals/connected` deep link the backend has stored the tokens,
 * so the app refreshes status and triggers a sync.
 */
class IntervalsService(private val api: NorthAxApi) {

    var connectionState by mutableStateOf<IntervalsConnectionState>(IntervalsConnectionState.Disconnected)
    var syncedActivities by mutableStateOf<List<GarminActivity>>(emptyList())
    var isSyncing by mutableStateOf(false)
        private set

    // MARK: - Status

    /** Load the current connection state from the backend (call on app start). */
    suspend fun refreshStatus() {
        try {
            val state = api.intervalsStatus()
            connectionState = state
            if (state.isConnected) {
                syncedActivities = try {
                    api.activities(limit = 30)
                } catch (_: Exception) {
                    syncedActivities
                }
            }
        } catch (_: Exception) {
        }
    }

    // MARK: - Connection

    /** Start the OAuth flow: returns the authorization URL to open in a Custom
     *  Tab, or null on failure (connectionState carries the error). */
    suspend fun beginConnect(): String? {
        connectionState = IntervalsConnectionState.Connecting
        return try {
            api.intervalsAuthorizationUrl()
        } catch (e: ApiError) {
            connectionState = if (e.isIntervalsNotConfigured) {
                IntervalsConnectionState.Error("intervals.icu isn't enabled on this server yet. Please contact support.")
            } else {
                IntervalsConnectionState.Error(e.userMessage)
            }
            null
        } catch (_: Exception) {
            connectionState = IntervalsConnectionState.Error("Couldn't connect to intervals.icu.")
            null
        }
    }

    /** Called when the `northax://intervals/connected` deep link arrives. */
    suspend fun completeConnect() {
        try {
            connectionState = api.intervalsStatus()
            syncActivities()
        } catch (_: Exception) {
            connectionState = IntervalsConnectionState.Error("Couldn't connect to intervals.icu.")
        }
    }

    /** Connect with a personal intervals.icu API key (no web OAuth). */
    suspend fun connectWithApiKey(athleteId: String, apiKey: String) {
        connectionState = IntervalsConnectionState.Connecting
        try {
            connectionState = api.connectWithApiKey(athleteId, apiKey)
            syncActivities()
        } catch (e: ApiError) {
            connectionState = IntervalsConnectionState.Error(e.userMessage)
        } catch (_: Exception) {
            connectionState = IntervalsConnectionState.Error("Couldn't connect with that API key.")
        }
    }

    suspend fun disconnect() {
        try {
            api.intervalsDisconnect()
        } catch (_: Exception) {
        }
        connectionState = IntervalsConnectionState.Disconnected
        syncedActivities = emptyList()
    }

    // MARK: - Sync

    suspend fun syncActivities() {
        if (!connectionState.isConnected) return
        isSyncing = true
        try {
            api.intervalsSync()
            syncedActivities = api.activities(limit = 30)
            connectionState = api.intervalsStatus()
        } catch (_: Exception) {
            // Keep the last good state; a transient sync failure isn't fatal.
        } finally {
            isSyncing = false
        }
    }

    // MARK: - Plan push

    /** Push a planned session to intervals.icu (→ Garmin) as a scheduled
     *  workout. Returns whether the push succeeded. */
    suspend fun pushPlannedSession(session: PlannedSession, date: LocalDate): Boolean {
        if (!connectionState.isConnected) return false
        return try {
            api.pushWorkout(date, session)
            true
        } catch (_: Exception) {
            false
        }
    }
}
